src/realm/CMakeLists.txt

```diff
@@ -0,0 +1,324 @@
+    util/to_string.hpp
+    util/tuple.hpp
+    util/type_list.hpp
+    util/type_traits.hpp
+    util/utf8.hpp
+) # REALM_INSTALL_UTIL_HEADERS
+
+set(REALM_OTHER_UTIL_HEADERS
+    util/aes_cryptor.hppd
+    util/errno.hpp
+) # REALM_OTHER_UTIL_HEADERS
+
+add_library(realm-objects OBJECT ${REALM_SOURCES})
+set_target_properties(realm-objects PROPERTIES POSITION_INDEPENDENT_CODE true)
+target_compile_definitions(realm-objects PUBLIC PIC)
+
+if(APPLE)
+    set_target_properties(
+            realm-objects PROPERTIES
+            XCODE_ATTRIBUTE_SKIP_INSTALL "NO"
+            XCODE_ATTRIBUTE_GCC_OPTIMIZATION_LEVEL "\$(GCC_OPTIMIZATION_LEVEL_\$(CONFIGURATION))"
+            XCODE_ATTRIBUTE_GCC_OPTIMIZATION_LEVEL_Debug "0"
+            XCODE_ATTRIBUTE_GCC_OPTIMIZATION_LEVEL_MinSizeDebug "z"
+            XCODE_ATTRIBUTE_GCC_OPTIMIZATION_LEVEL_Release "3"
+            XCODE_ATTRIBUTE_GCC_OPTIMIZATION_LEVEL_RelWithDebInfo "3"
+            XCODE_ATTRIBUTE_GCC_OPTIMIZATION_LEVEL_RelMinSize "3"
+    )
+endif()
+
+if(CMAKE_SYSTEM_NAME MATCHES "^Windows")
+    add_dependencies(realm-objects pthread_win32 getopt_win32)
+    set_property(TARGET realm-objects APPEND PROPERTY COMPILE_DEFINITIONS PTW32_STATIC_LIB)
+    include_directories(realm-objects INTERFACE
+                               ${CMAKE_CURRENT_SOURCE_DIR}/../win32
+                               ${CMAKE_CURRENT_SOURCE_DIR}/../win32/pthread)
+    list(APPEND REALM_EXTRA_LIBS pthread_win32 getopt_win32)
+endif()
+
+# The placeholder file is needed in order to make builds work on Xcode
+# The issue is described in the last sentence here:
+# https://cmake.org/cmake/help/v3.7/command/add_library.html#object-libraries
+add_library(realm STATIC $<TARGET_OBJECTS:realm-objects> placeholder.cpp)
+
+target_link_libraries(
+        realm
+        ${REALM_EXTRA_LIBS}
+        ${CMAKE_THREAD_LIBS_INIT}
+)
+
+if(ANDROID)
+  target_link_libraries(realm atomic)
+endif()
+
+if(ANDROID OR CMAKE_SYSTEM_NAME MATCHES "^Windows")
+    set(REALM_SKIP_SHARED_LIB ON)
+endif()
+
+if(NOT REALM_SKIP_SHARED_LIB)
+    add_library(realm-shared SHARED $<TARGET_OBJECTS:realm-objects> placeholder.cpp)
+
+    set_target_properties(
+        realm-shared PROPERTIES
+        LIBRARY_OUTPUT_NAME realm
+    )
+
+    target_link_libraries(
+        realm-shared
+        ${REALM_EXTRA_LIBS}
+        ${CMAKE_THREAD_LIBS_INIT}
+    )
+
+  if(ANDROID)
+    target_link_libraries(realm-shared atomic)
+  endif()
+endif()
+
+if(REALM_ENABLE_ENCRYPTION AND NOT APPLE)
+    if(ANDROID)
+        string(TOLOWER "${CMAKE_BUILD_TYPE}" BUILD_TYPE)
+        set(OPENSSL_FILENAME "openssl-${BUILD_TYPE}-${ANDROID_OPENSSL_VERSION}-Android-${ANDROID_ABI}")
+        set(OPENSSL_URL  "http://static.realm.io/downloads/openssl/${ANDROID_OPENSSL_VERSION}/Android/${ANDROID_ABI}/${OPENSSL_FILENAME}.tar.gz")
+
+        message(STATUS "Downloading OpenSSL...")
+        file(DOWNLOAD "${OPENSSL_URL}" "${CMAKE_BINARY_DIR}/${OPENSSL_FILENAME}.tar.gz")
```

Comment:
I didn't understand that this was downloading a prebuilt binary, so that is awesome.

Would it be worth verifying a SHA1 of the downloaded tgz just in case?